test_collections/sdk_tests/support/python_testing/models/test_case.py

```diff
@@ -0,0 +1,281 @@
+        """Common class factory method for all subclasses of PythonTestCase."""
+        identifier = cls.__test_identifier(test.name)
+        class_name = cls.__class_name(identifier)
+        title = identifier
+
+        return type(
+            class_name,
+            (cls,),
+            {
+                "python_test": test,
+                "python_test_version": python_test_version,
+                "chip_tool_test_identifier": class_name,
+                "metadata": {
+                    "public_id": identifier,
+                    "version": "0.0.1",
+                    "title": title,
+                    "description": test.name,
+                },
+            },
+        )
+
+    @staticmethod
+    def __test_identifier(name: str) -> str:
+        """Find TC-XX-1.1 in Python Test title.
+        Note some have [TC-XX-1.1] and others TC-XX-1.1
+        """
+        title_pattern = re.compile(r"(?P<title>TC-[^\s\]]*)")
+        if match := re.search(title_pattern, name):
+            return match["title"]
+        else:
+            return name
+
+    @staticmethod
+    def __class_name(identifier: str) -> str:
+        """Replace all non-alphanumeric characters with _ to make valid class name."""
+        return re.sub("[^0-9a-zA-Z]+", "_", identifier)
+
+    def run_command(self, cmd: str) -> None:
+        os.system(cmd)
+
+    async def execute(self) -> None:
+        try:
+            logger.info("Running Python Test: " + self.metadata["title"])
+            BaseManager.register("TestRunnerHooks", SDKPythonTestRunnerHooks)
+            manager = BaseManager(address=("0.0.0.0", 50000), authkey=b"abc")
+            manager.start()
+            test_runner_hooks = manager.TestRunnerHooks()  # type: ignore
+            runner_class = RUNNER_CLASS_PATH + RUNNER_CLASS
+            self.chip_tool.send_command(
+                f"{runner_class} {self.metadata['title']}", prefix=EXECUTABLE
+            )
+            while ((update := test_runner_hooks.updates_test()) is not None) or (
+                not test_runner_hooks.finished()
+            ):
+                if not update:
+                    continue
+
+                def handle_update(update: dict) -> None:
+                    def call_function(obj, func_name, kwargs) -> None:  # type: ignore
+                        func = getattr(obj, func_name, None)
+                        if not func:
+                            raise AttributeError(
+                                f"{func_name} is not a method of {obj}"
+                            )
+                        if not callable(func):
+                            raise TypeError(f"{func_name} is not callable")
+                        # Call the method with the unpacked keyword arguments.
+                        func(**kwargs)
+
+                    for func_name, kwargs in update.items():
+                        call_function(self, func_name, kwargs)
+
+                handle_update(update)
```

Comment:
Could this `call_function` be defined somewhere else? I think it would be more readable if all this was reduced to just:

```python
for func_name, kwargs in update.items():
    call_function(self, func_name, kwargs)
```

...and the `call_function` method be defined outside of `execute`.